### PR TITLE
Add game state and and Player name to Game Document

### PR DESCRIPTION
**Pull request by barmettlerl**

### CODE REVIEW

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt [hunk 1]
```diff
@@ -199,7 +208,28 @@ private data class StoreMatch(
...
+
+    store.collection("games").document(id).update {
+      this[FieldPath(listOf("metadata", "status"))] =
+          when (game.nextStep) {
+            NextStep.Stalemate -> "stalemate"
+            is NextStep.MovePiece -> null
+            else ->
+                if ((game.nextStep as NextStep.Checkmate).winner == Color.Black) "blackWon"
+                else "whiteWon"
+          }
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt:221]: What you're looking for is this syntax :

```kotlin
when (val step = game.nextStep) {
  NextStep.Stalemate -> TODO()
  is NextStep.MovePiece -> TODO()
  is NextStep.Checkmate -> // here you can use step.winner, it's smart-casted.
}
```

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt [hunk 2]
```diff
@@ -199,7 +208,28 @@ private data class StoreMatch(
       }
 
   override suspend fun update(game: Game) {
-    store.collection("games").document(id).update { this["moves"] = game.toAlgebraicNotation() }
+    val chess = store.collection("games").document(id).get<ChessDocument>()
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt:211]: I'm not sure `chess` is the most suitable name for this. What about `document` for instance ?

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt [hunk 3]
```diff
@@ -199,7 +208,28 @@ private data class StoreMatch(
...
+      this[FieldPath(listOf("metadata", "status"))] =
+          when (game.nextStep) {
+            NextStep.Stalemate -> "stalemate"
+            is NextStep.MovePiece -> null
+            else ->
+                if ((game.nextStep as NextStep.Checkmate).winner == Color.Black) "blackWon"
+                else "whiteWon"
+          }
+
+      if (chess?.blackId == user?.uid && chess?.metadata?.blackName != user?.name) {
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt:223]: Note : updating a field with the same value is not something we have to actively prevent imho

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt [hunk 4]
```diff
@@ -81,21 +83,27 @@ class ChessFacade(
...
   }
 
-  private fun getMatchesForPlayer(colorField: String, playerId: String): Flow<List<Match>> {
-    return store.collection("games").whereEquals(colorField, playerId).asMatchListFlow().onStart {
-      emit(emptyList())
-    }
+  private fun getMatchesForPlayer(
+      colorField: String,
+      playerId: String,
+      user: Profile? = null
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt:95]: What would you think of making the `user` argument mandatory and using `user.uid` to replace `playerId` ?

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/StoreDocuments.kt [hunk 5]
```diff
@@ -64,6 +64,14 @@ data class ChessDocument(
     val moves: List<String>? = null,
     val whiteId: String? = null,
     val blackId: String? = null,
+    val lastUpdatedAt: Long? = null,
+    val metadata: ChessMetadata? = null
+)
+
+data class ChessMetadata(
+    val status: String? = null, // whiteWon, blackWon, draw
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/application/StoreDocuments.kt:72]: It's worth specifying these magic values as constants, maybe in a `companion object` for `ChessMetadata`

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/StoreDocuments.kt [hunk 6]
```diff
@@ -64,6 +64,14 @@ data class ChessDocument(
     val moves: List<String>? = null,
     val whiteId: String? = null,
     val blackId: String? = null,
+    val lastUpdatedAt: Long? = null,
+    val metadata: ChessMetadata? = null
+)
+
+data class ChessMetadata(
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/application/StoreDocuments.kt:71]: Missing some documentation.

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/StoreDocuments.kt [hunk 7]
```diff
@@ -64,6 +64,14 @@ data class ChessDocument(
     val moves: List<String>? = null,
     val whiteId: String? = null,
     val blackId: String? = null,
+    val lastUpdatedAt: Long? = null,
+    val metadata: ChessMetadata? = null
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/application/StoreDocuments.kt:68]: Missing documentation for these two new fields.

File: mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulGameScreen.kt [hunk 8]
```diff
@@ -56,7 +56,7 @@ fun StatefulGameScreen(
   val speechFacade = LocalSpeechFacade.current
 
   val scope = rememberCoroutineScope()
-  val match = remember(chessFacade, id) { chessFacade.match(id) }
+  val match = remember(chessFacade, id) { chessFacade.match(id, user) }
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulGameScreen.kt:59]:
```suggestion
  val match = remember(chessFacade, id, user) { chessFacade.match(id, user) }
```

File: mobile/src/androidTest/java/ch/epfl/sdp/mobile/test/application/ChessFacadeTest.kt [hunk 9]
```diff
@@ -136,4 +141,136 @@ class ChessFacadeTest {
     assertThat(newSolvedPuzzles.size).isEqualTo(1)
     assertThat(newUnsolvedPuzzles).isEmpty()
   }
+
+  @Test
+  fun given_userChanngedName_when_aGameMovehasBeenMade_then_whiteNameInMetadataShouldUpdate() =
```
Comment (alexandrepiveteau) [inline mobile/src/androidTest/java/ch/epfl/sdp/mobile/test/application/ChessFacadeTest.kt:146]: Typo

```suggestion
  fun given_userChangedName_when_aGameMoveHasBeenMade_then_whiteNameInMetadataShouldUpdate() =
```

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt [hunk 10]
```diff
@@ -199,7 +208,28 @@ private data class StoreMatch(
...
+          }
+
+      if (chess?.blackId == user?.uid && chess?.metadata?.blackName != user?.name) {
+        this[FieldPath(listOf("metadata", "blackName"))] = user?.name
+      }
+
+      if (chess?.whiteId == user?.uid && chess?.metadata?.whiteName != user?.name) {
+        this[FieldPath(listOf("metadata", "whiteName"))] = user?.name
+      }
+
```
Comment (alexandrepiveteau) [inline mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt:230]: I think you should be updating the last updated timestamp for the chess document here.

**Review by alexandrepiveteau — COMMENTED** [review]
Nice work !

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt (hunk 1 above)
Comment (barmettlerl) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt:221]: thx, much more cleaner

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt (hunk 3 above)
Comment (barmettlerl) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt:223]: alright

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt (hunk 4 above)
Comment (barmettlerl) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt:95]: Ah thx a lot cleaner

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/StoreDocuments.kt (hunk 5 above)
Comment (barmettlerl) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/StoreDocuments.kt:72]: yea I think this is better. It felt a bit like random strings in the code

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/StoreDocuments.kt (hunk 6 above)
Comment (barmettlerl) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/StoreDocuments.kt:71]: yes

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/StoreDocuments.kt (hunk 7 above)
Comment (barmettlerl) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/StoreDocuments.kt:68]: yes

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt (hunk 10 above)
Comment (barmettlerl) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt:233]: yes

File: mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt (hunk 2 above)
Comment (barmettlerl) [reply mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/ChessFacade.kt:214]: is done

**Review by matt989253 — APPROVED** [review]

**Review by matt989253 — APPROVED** [review]
👍🏻

**Pull request merged**